File: python3/day08/day08.py
from copy import deepcopy


original_program = [l.split() for l in open("input.txt").readlines()]
program_length = len(original_program)


def execute_instruction(program, ip, acc):
    instruction, operand = program[ip]

    if instruction == "acc":
        acc += int(operand)
        ip += 1

    elif instruction == "jmp":
        ip += int(operand)

    elif instruction == "nop":
        ip += 1

    return (ip, acc)


def run_program(program):
    ip = 0
    acc = 0
    already_executed = []
    while True:
        if ip in already_executed or ip == program_length:
            return ip, acc

        already_executed.append(ip)
        ip, acc = execute_instruction(program, ip, acc)


def part1():
    _, acc = run_program(original_program)
    return acc


def part2():
    for ip_to_change in range(program_length):
        if original_program[ip_to_change][0] == "jmp":
            modified_program = deepcopy(original_program)
            modified_program[ip_to_change][0] = "nop"
        elif (
            original_program[ip_to_change][0] == "nop"
            and original_program[ip_to_change][1] != "0"
        ):
            modified_program = deepcopy(original_program)
            modified_program[ip_to_change][0] = "jmp"
        else:
            continue

        ip, acc = run_program(modified_program)
        if ip == program_length:
            return acc


print(part1())
print(part2())

File: scala/src/main/scala/year2020/Day08.scala
package com.ochronus
package year2020

import utils.File

class Day08 {
  private val originalProgram = File.readFileLines("day08.input.txt").map(line => (line.split(' ')(0), line.split(' ')(1).toInt))

  def execute(ip: Int = 0, acc: Int = 0, alreadyExecuted: Set[Int] = Set.empty)
             (implicit program: Seq[(String, Int)]): (Int, Boolean) = {

    if (ip == program.size) {
      (acc, true)
    }
    else if (alreadyExecuted.contains(ip)) {
      (acc, false)
    } else {
      program(ip) match {
        case ("acc", value) => execute(ip + 1, acc + value, alreadyExecuted + ip)
        case ("jmp", value) => execute(ip + value, acc, alreadyExecuted + ip)
        case ("nop", _) => execute(ip + 1, acc, alreadyExecuted + ip)
      }
    }
  }

  def part1(): Long = {
    val res = execute()(originalProgram)
    res._1
  }

  def part2(): Option[Long] = {
    for ((instruction, ip) <- originalProgram.zipWithIndex) {
      val newInstruction = instruction match {
        case ("jmp", value) => ("nop", value)
        case ("nop", value) => if (value != 0) ("jmp", value) else ("nop", value)
        case ("acc", value) => ("acc", value)
      }
      if (newInstruction != instruction) execute()(originalProgram.updated(ip, newInstruction)) match {
        case (acc, true) => return Some(acc)
        case _ => None
      }
    }
    None
  }
}
